### PR TITLE
use gather_nd to _gather_states in LSTMBlockWapper

### PR DESCRIPTION
**Pull request by halleywj**
no need for calculate mod_indices and reshape data .

### CODE REVIEW

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py [hunk 1]
```diff
@@ -2771,7 +2771,9 @@ def call(self, inputs, state):
     c = f * state + (1.0 - f) * x_bar
     h = r * self._activation(c) + (1.0 - r) * inputs
 
-    return h, c
+    new_state = (rnn_cell_impl.LSTMStateTuple(c, h))
```
Comment (drpngx) [inline tensorflow/contrib/rnn/python/ops/rnn_cell.py]: why this change?

File: tensorflow/contrib/rnn/python/ops/lstm_ops.py [hunk 2]
```diff
@@ -572,9 +572,8 @@ def call(self, inputs, initial_state=None, dtype=None, sequence_length=None):
 
   def _gather_states(self, data, indices, batch_size):
     """Produce `out`, s.t. out(i, j) = data(indices(i), i, j)."""
-    mod_indices = indices * batch_size + math_ops.range(batch_size)
-    return array_ops.gather(
-        array_ops.reshape(data, [-1, self.num_units]), mod_indices)
+    return array_ops.gather_nd(
```
Comment (drpngx) [inline tensorflow/contrib/rnn/python/ops/lstm_ops.py]: What's the motivation for this?

**Review by drpngx — CHANGES_REQUESTED** [review]

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py (hunk 1 above)
Comment (halleywj) [reply tensorflow/contrib/rnn/python/ops/rnn_cell.py]: Because many other cells , such as BasicLSTMCell 、 LSTMCell 、NASCell 、 LayerNormBasicLSTMCell , return h and stateTuple with c and h when call . I think sru should return the same type as lstm .

File: tensorflow/contrib/rnn/python/ops/lstm_ops.py (hunk 2 above)
Comment (halleywj) [reply tensorflow/contrib/rnn/python/ops/lstm_ops.py]: Because _gather_states find out(i, j) = data(indices(i), i, j) when indices is a [batch_size] tensor . The origin method  reshape the 'data' to a 2-dim tensor and calculate the indices in the reshaped array . Use gather_nd can locate the 'indices' in the 3-dim tensor 'data' so no need for reshape . And in my experiment , gather may lead to a warning when data is a sparse tensor but use gather_nd with no warning . And use gather_nd can speed up a little .

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py (hunk 1 above)
Comment (drpngx) [reply tensorflow/contrib/rnn/python/ops/rnn_cell.py]: So I think that should be split into another PR to keep things clean.

File: tensorflow/contrib/rnn/python/ops/lstm_ops.py (hunk 2 above)
Comment (drpngx) [reply tensorflow/contrib/rnn/python/ops/lstm_ops.py]: OK, makes sense.

**Review by drpngx — CHANGES_REQUESTED** [review]

File: tensorflow/contrib/rnn/python/ops/rnn_cell.py (hunk 1 above)
Comment (halleywj) [reply tensorflow/contrib/rnn/python/ops/rnn_cell.py]: OK，I will do that.

**Comment by halleywj** [comment]
@drpngx I have clean my commit .

**Review by drpngx — APPROVED** [review]
